engine/apps/shift_swaps/models/shift_swap_request.py

```diff
@@ -0,0 +1,151 @@
+import typing
+
+from django.conf import settings
+from django.core.validators import MinLengthValidator
+from django.db import models
+from django.utils import timezone
+
+from apps.shift_swaps import exceptions
+from common.public_primary_keys import generate_public_primary_key, increase_public_primary_key_length
+
+if typing.TYPE_CHECKING:
+    from apps.user_management.models import User
+
+
+def generate_public_primary_key_for_shift_swap_request() -> str:
+    prefix = "SSR"
+    new_public_primary_key = generate_public_primary_key(prefix)
+
+    failure_counter = 0
+    while ShiftSwapRequest.objects.filter(public_primary_key=new_public_primary_key).exists():
+        new_public_primary_key = increase_public_primary_key_length(
+            failure_counter=failure_counter, prefix=prefix, model_name="ShiftSwapRequest"
+        )
+        failure_counter += 1
+
+    return new_public_primary_key
+
+
+class ShiftSwapRequestQueryset(models.QuerySet):
+    def delete(self):
+        self.update(deleted_at=timezone.now())
+
+
+class ShiftSwapRequestManager(models.Manager):
+    def get_queryset(self):
+        return ShiftSwapRequestQueryset(self.model, using=self._db).filter(deleted_at=None)
+
+    def hard_delete(self):
+        return self.get_queryset().hard_delete()
+
+
+class ShiftSwapRequest(models.Model):
+
+    objects = ShiftSwapRequestManager()
+    objects_with_deleted = models.Manager()
+
+    public_primary_key = models.CharField(
+        max_length=20,
+        validators=[MinLengthValidator(settings.PUBLIC_PRIMARY_KEY_MIN_LENGTH + 1)],
+        unique=True,
+        default=generate_public_primary_key_for_shift_swap_request,
+    )
+
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+    deleted_at = models.DateTimeField(null=True)
+
+    schedule = models.ForeignKey(
+        to="schedules.OnCallSchedule", null=False, on_delete=models.CASCADE, related_name="shift_swap_requests"
+    )
+
+    swap_start = models.DateTimeField()
+    """
+    so long as objects are created through the internal API, `swap_start` is guaranteed to be in UTC
+    (see the internal API serializer for more details)
+    """
+
+    swap_end = models.DateTimeField()
+    """
+    so long as objects are created through the internal API, `swap_end` is guaranteed to be in UTC
+    (see the internal API serializer for more details)
+    """
+
+    description = models.TextField(max_length=3000, default=None, null=True)
+
+    beneficiary = models.ForeignKey(
+        to="user_management.User", null=False, on_delete=models.CASCADE, related_name="created_shift_swap_requests"
+    )
+    """
+    the person who is relieved from (part of) their shift(s)
+    """
+
+    benefactor = models.ForeignKey(
+        to="user_management.User", null=True, on_delete=models.CASCADE, related_name="taken_shift_swap_requests"
+    )
+    """
+    the person taking on shift workload from the beneficiary
+    """
```

Comment:
these docstrings may look a bit odd, but they follow [PEP-224](https://peps.python.org/pep-0224/) (Attribute Docstrings). VS Code gives you a nice tooltip w/ this info:
<img width="757" alt="Screenshot 2023-07-20 at 23 44 38" src="https://github.com/grafana/oncall/assets/9406895/5a3eaed5-b9ee-4d88-ba7f-1c4525f6c472">